README.md
# wywygmbh/log4j-plugin-fluency log4j to fluentd log appender
[![GitHub version](https://badge.fury.io/gh/wywygmbh%2Flog4j-plugin-fluency.svg)](https://badge.fury.io/gh/wywygmbh%2Flog4j-plugin-fluency)
[![Build Status](https://travis-ci.org/wywygmbh/log4j-plugin-fluency.svg?branch=master)](https://travis-ci.org/wywygmbh/log4j-plugin-fluency)

Simple logging from java to [fluentd](http://www.fluentd.org/) using Log4j 2

## Features

This is a thin wrapper taking Log4j log events, adding some metadata and passing them to fluency

* easy to integrate Log4j 2 appender (no code changes required!)
* fluentd forward protocol
* all the benefits from [fluency](https://github.com/komamitsu/fluency)

<img src="screenshot.png" width="600">

## Install

Place the `.jar` in your classpath or using maven:

```xml
<dependency>
  <groupId>com.wywy</groupId>
  <artifactId>log4j-plugin-fluency</artifactId>
  <version>1.3.2</version>
</dependency>
```

## Usage

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Configuration packages="com.wywy.log4j.appender">
    <Appenders>
        <Fluency name="fluency" tag="yourTag">
            <!-- all settings are optional, see defaultFluency() for default values -->
            <!-- you can add as may fields as you like (or none at all) -->
            <StaticField name="application">yourApplication</StaticField>
            <StaticField name="someOtherField">some value</StaticField>
            <Server host="primary-node" port="24224"/>
            <Server host="secondary-node" port="24224"/>
            <FluencyConfig
              ackResponseMode="true"
              fileBackupDir="/tmp/fluency"
              bufferChunkInitialSize="4194304"
              bufferChunkRetentionSize="16777216"
              maxBufferSize="268435456"
              waitUntilBufferFlushed="30"
              waitUntilFlusherTerminated="40"
              flushIntervalMillis="200"
              senderMaxRetryCount="12" />
        </Fluency>
    </Appenders>
    <Loggers>
        <root level="info"> <!-- We log everything -->
            <appender-ref ref="fluency"/>
        </root>
    </Loggers>
</Configuration>
```

`packages` is required for log4j to find the Appender ([see](https://logging.apache.org/log4j/2.x/manual/configuration.html#ConfigurationSyntax))

### Configuration

`tag`

  **required** - tag to identify messages, can be used in fluentd for message manipulation

## Internals

The appender gets initialized by log4j and creates a new `Fluency.defaultFluency();`.

All log messages get aggregated with some additional fields and handed over to fluency.

fluency will forward the messages to a fluentd running on localhost.

## License

Copyright 2017 wywy GmbH

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

This code is being actively maintained by some fellow engineers at [wywy GmbH](http://wywy.com/).


CHANGELOG.md
# Changelog

## 1.3.2

### Fixes

* handle log4j 2.6 signature change - thanks [@askannon](https://github.com/askannon)

## 1.3.1

### Fixes

* fix log messages with exceptions - they were dropped and only the message got logged

## 1.3.0

### Features

* [[#2](https://github.com/wywygmbh/log4j-plugin-fluency/pull/2)] add support for fluency configuration options - thanks [@takumakanari](https://github.com/takumakanari)

## 1.2.0

### Features

* [[#1](https://github.com/wywygmbh/log4j-plugin-fluency/pull/1)] add support for PatternLayout - thanks [@takumakanari](https://github.com/takumakanari)

## 1.1.1
* switch to `StatusLogger` for internal logging

## 1.1.0
* add generic `<StaticField>` configuration
    * deprecates `application=""` attribute (it still works, but please migrate)

## 1.0.0
* switch groupId for maven central registration

## 0.1.0
* initial release


src/com/wywy/log4j/appender/Server.java
package com.wywy.log4j.appender;

import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.core.config.Node;
import org.apache.logging.log4j.core.config.plugins.Plugin;
import org.apache.logging.log4j.core.config.plugins.PluginAttribute;
import org.apache.logging.log4j.core.config.plugins.PluginFactory;
import org.apache.logging.log4j.status.StatusLogger;

import java.net.InetSocketAddress;

/**
 * Represents a host/port pair in the configuration.
 */
@Plugin(name = "Server", category = Node.CATEGORY, printObject = true)
public class Server {

    private static final Logger LOGGER = StatusLogger.getLogger();

    private final String host;
    private final int port;

    private Server(final String host, final int port) {
        this.host = host;
        this.port = port;
    }

    public InetSocketAddress configure() {
        return new InetSocketAddress(host, port);
    }

    @PluginFactory
    public static Server createServer(@PluginAttribute("host") final String host,
            @PluginAttribute("port") final int port) {
        if (host == null) {
            LOGGER.error("Property host cannot be null");
        }
        if (port <= 0) {
            LOGGER.error("Property port must be > 0");
        }
        return new Server(host, port);
    }

    @Override
    public String toString() {
        return "Server{" +
                "host='" + host + '\'' +
                ", port=" + port +
                '}';
    }
}


src/com/wywy/log4j/appender/StaticField.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache license, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the license for the specific language governing permissions and
 * limitations under the license.
 */
/*
 * this file is copied from Property.java in log4j - but the name Property would look weird in the xml, so it's renamed here
 * https://github.com/apache/logging-log4j2/blob/master/log4j-core/src/main/java/org/apache/logging/log4j/core/config/Property.java
 */
package com.wywy.log4j.appender;

import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.core.config.Node;
import org.apache.logging.log4j.core.config.plugins.Plugin;
import org.apache.logging.log4j.core.config.plugins.PluginAttribute;
import org.apache.logging.log4j.core.config.plugins.PluginFactory;
import org.apache.logging.log4j.core.config.plugins.PluginValue;
import org.apache.logging.log4j.status.StatusLogger;

/**
 * Represents a key/value pair in the configuration.
 */
@Plugin(name = "StaticField", category = Node.CATEGORY, printObject = true)
public final class StaticField {

    private static final Logger LOGGER = StatusLogger.getLogger();

    private final String name;
    private final String value;
    private final boolean valueNeedsLookup;

    private StaticField(final String name, final String value) {
        this.name = name;
        this.value = value;
        this.valueNeedsLookup = value != null && value.contains("${");
    }

    /**
     * Returns the property name.
     * @return the property name.
     */
    public String getName() {
        return name;
    }

    /**
     * Returns the property value.
     * @return the value of the property.
     */
    public String getValue() {
        return value == null ? "" : value; // LOG4J2-1313 null would be same as Property not existing
    }

    /**
     * Returns {@code true} if the value contains a substitutable property that requires a lookup to be resolved.
     * @return {@code true} if the value contains {@code "$&#123;"}, {@code false} otherwise
     */
    public boolean isValueNeedsLookup() {
        return valueNeedsLookup;
    }

    /**
     * Creates a Property.
     *
     * @param name The key.
     * @param value The value.
     * @return A Property.
     */
    @PluginFactory
    public static StaticField createStaticField(
            @PluginAttribute("name") final String name,
            @PluginValue("value") final String value) {
        if (name == null) {
            LOGGER.error("Property name cannot be null");
        }
        return new StaticField(name, value);
    }

    @Override
    public String toString() {
        return name + '=' + getValue();
    }
}


src/com/wywy/log4j/appender/FluencyConfig.java
package com.wywy.log4j.appender;

import org.apache.logging.log4j.core.config.Node;
import org.apache.logging.log4j.core.config.plugins.Plugin;
import org.apache.logging.log4j.core.config.plugins.PluginAttribute;
import org.apache.logging.log4j.core.config.plugins.PluginFactory;
import org.komamitsu.fluency.Fluency;

/**
 * Represents Fluency params in the configuration.
 */
@Plugin(name = "FluencyConfig", category = Node.CATEGORY, printObject = true)
public class FluencyConfig {

    private boolean ackResponseMode;
    private String fileBackupDir;
    private int bufferChunkInitialSize;
    private int bufferChunkRetentionSize;
    private Long maxBufferSize;
    private int waitUntilBufferFlushed;
    private int waitUntilFlusherTerminated;
    private int flushIntervalMillis;
    private int senderMaxRetryCount;

    private FluencyConfig() {
    }

    public Fluency.Config configure() {
        Fluency.Config config = new Fluency.Config().setAckResponseMode(ackResponseMode);
        if (fileBackupDir != null) config.setFileBackupDir(fileBackupDir);
        if (bufferChunkInitialSize > 0) config.setBufferChunkInitialSize(bufferChunkInitialSize);
        if (bufferChunkRetentionSize > 0) config.setBufferChunkRetentionSize(bufferChunkRetentionSize);
        if (maxBufferSize > 0) config.setMaxBufferSize(maxBufferSize);
        if (waitUntilBufferFlushed > 0) config.setWaitUntilBufferFlushed(waitUntilBufferFlushed);
        if (waitUntilFlusherTerminated > 0) config.setWaitUntilFlusherTerminated(waitUntilFlusherTerminated);
        if (flushIntervalMillis > 0) config.setFlushIntervalMillis(flushIntervalMillis);
        if (senderMaxRetryCount > 0) config.setSenderMaxRetryCount(senderMaxRetryCount);
        return config;
    }

    @PluginFactory
    public static FluencyConfig createFluencyConfig(
            @PluginAttribute("ackResponseMode") final boolean ackResponseMode,
            @PluginAttribute("fileBackupDir") final String fileBackupDir,
            @PluginAttribute("bufferChunkInitialSize") final int bufferChunkInitialSize,
            @PluginAttribute("bufferChunkRetentionSize") final int bufferChunkRetentionSize,
            @PluginAttribute("maxBufferSize") final Long maxBufferSize,
            @PluginAttribute("waitUntilBufferFlushed") final int waitUntilBufferFlushed,
            @PluginAttribute("waitUntilFlusherTerminated") final int waitUntilFlusherTerminated,
            @PluginAttribute("flushIntervalMillis") final int flushIntervalMillis,
            @PluginAttribute("senderMaxRetryCount") final int senderMaxRetryCount) {
        FluencyConfig config = new FluencyConfig();
        config.ackResponseMode = ackResponseMode;
        config.fileBackupDir = fileBackupDir;
        config.bufferChunkInitialSize = bufferChunkInitialSize;
        config.bufferChunkRetentionSize = bufferChunkRetentionSize;
        config.maxBufferSize = maxBufferSize;
        config.waitUntilBufferFlushed = waitUntilBufferFlushed;
        config.waitUntilFlusherTerminated = waitUntilFlusherTerminated;
        config.flushIntervalMillis = flushIntervalMillis;
        config.senderMaxRetryCount = senderMaxRetryCount;
        return config;
    }

    @Override
    public String toString() {
        return "FluencyConfig{" +
                "ackResponseMode=" + ackResponseMode +
                ", fileBackupDir='" + fileBackupDir + '\'' +
                ", bufferChunkInitialSize=" + bufferChunkInitialSize +
                ", bufferChunkRetentionSize=" + bufferChunkRetentionSize +
                ", maxBufferSize=" + maxBufferSize +
                ", waitUntilBufferFlushed=" + waitUntilBufferFlushed +
                ", waitUntilFlusherTerminated=" + waitUntilFlusherTerminated +
                ", flushIntervalMillis=" + flushIntervalMillis +
                ", senderMaxRetryCount=" + senderMaxRetryCount +
                '}';
    }
}


src/com/wywy/log4j/appender/FluencyAppender.java
package com.wywy.log4j.appender;


import org.apache.logging.log4j.core.*;
import org.apache.logging.log4j.core.appender.AbstractAppender;
import org.apache.logging.log4j.core.config.plugins.Plugin;
import org.apache.logging.log4j.core.config.plugins.PluginAttribute;
import org.apache.logging.log4j.core.config.plugins.PluginElement;
import org.apache.logging.log4j.core.config.plugins.PluginFactory;
import org.apache.logging.log4j.core.layout.PatternLayout;
import org.apache.logging.log4j.core.pattern.NameAbbreviator;
import org.apache.logging.log4j.core.util.Booleans;
import org.apache.logging.log4j.status.StatusLogger;
import org.komamitsu.fluency.Fluency;

import java.io.IOException;
import java.io.Serializable;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.InetSocketAddress;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Plugin(name="Fluency", category="Core", elementType="appender", printObject=true)
public final class FluencyAppender extends AbstractAppender {

    private static final StatusLogger LOG = StatusLogger.getLogger();
    private final SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
    private final NameAbbreviator abbr = NameAbbreviator.getAbbreviator("1.");
    private boolean usePre26Abbreviate = false;
    private Method abbreviateMethod;

    private Fluency fluency;
    private Map<String, Object> parameters;
    private Map<String, String> staticFields;

    private FluencyAppender(final String name, final Map<String, Object> parameters, final Map<String, String> staticFields,
                            final Server[] servers, final FluencyConfig fluencyConfig, final Filter filter,
                            final Layout<? extends Serializable> layout, final boolean ignoreExceptions) {

        super(name, filter, layout, ignoreExceptions);

        this.parameters = parameters;
        this.staticFields = staticFields;

        try {
            this.fluency = makeFluency(servers, fluencyConfig);
            LOG.info("FluencyAppender initialized");
        } catch (IOException e) {
            LOG.error(e.getMessage());
        }
        
        try {
            abbreviateMethod = NameAbbreviator.class.getMethod("abbreviate", new Class[] { String.class });
            usePre26Abbreviate = true;
        } catch (final NoSuchMethodException e) {
            try {
                abbreviateMethod = NameAbbreviator.class.getMethod("abbreviate", new Class[] { String.class, StringBuilder.class });
            } catch (final NoSuchMethodException | SecurityException e1) {
                LOG.error(e.getMessage(), e);
            }
        } catch (final SecurityException e) {
            LOG.error(e.getMessage(), e);
        }
        
    }

    @PluginFactory
    public static FluencyAppender createAppender(@PluginAttribute("name") final String name,
                                                 @PluginAttribute("tag") final String tag,
                                                 @PluginAttribute("application") final String application,
                                                 @PluginAttribute("ignoreExceptions") final String ignore,
                                                 @PluginElement("StaticField") final StaticField[] staticFields,
                                                 @PluginElement("Server") final Server[] servers,
                                                 @PluginElement("FluencyConfig") final FluencyConfig fluencyConfig,
                                                 @PluginElement("Layout") Layout<? extends Serializable> layout,
                                                 @PluginElement("Filter") final Filter filter) {
        final boolean ignoreExceptions = Booleans.parseBoolean(ignore, true);

        Map<String, Object> parameters = new HashMap<>();
        Map<String, String> fields = new HashMap<>();

        // the @Required annotation for Attributes just returns a non helpful exception, so i'll check it manually
        // we need the tag for fluency itself (it's actually a part of the fluentd protocol
        if (tag != null) {
            parameters.put("tag", tag);
        } else {
            throw new IllegalArgumentException("tag is required");
        }

        // Deprecated
        if (application != null) {
            fields.put("application", application);
        }

        for (StaticField field: staticFields) {
            if (field.getName().trim().equals("")) {
                LOG.warn("Skipping empty field");
                continue;
            }
            if (field.getValue().trim().equals("")) {
                LOG.warn("Skipping field {} due to empty value", field.getName());
                continue;
            }
            fields.put(field.getName(), field.getValue());
        }

        if (layout == null) {
            layout = PatternLayout.createDefaultLayout();
        }

        return new FluencyAppender(name, parameters, fields, servers, fluencyConfig, filter,
                layout, ignoreExceptions);
    }

    // These are the defaults, it no configuration is given:
    // Single Fluentd(localhost:24224 by default)
    //   - TCP heartbeat (by default)
    //   - Asynchronous flush (by default)
    //   - Without ack response (by default)
    //   - Flush interval is 600ms (by default)
    //   - Initial chunk buffer size is 1MB (by default)
    //   - Threshold chunk buffer size to flush is 4MB (by default)
    //   - Max total buffer size is 16MB (by default)
    //   - Max retry of sending events is 8 (by default)
    //   - Max wait until all buffers are flushed is 60 seconds (by default)
    //   - Max wait until the flusher is terminated is 60 seconds (by default)
    static Fluency makeFluency(Server[] servers, FluencyConfig config) throws IOException {
        if (servers.length == 0 && config == null) {
            return Fluency.defaultFluency();
        }
        if (servers.length == 0) {
            return Fluency.defaultFluency(config.configure());
        }
        List<InetSocketAddress> addresses = new ArrayList<>(servers.length);
        for (Server s : servers) {
            addresses.add(s.configure());
        }
        if (config == null) {
            return Fluency.defaultFluency(addresses);
        }
        return Fluency.defaultFluency(addresses, config.configure());
    }

    @Override
    public void append(LogEvent logEvent) {
        String level = logEvent.getLevel().name();
        String loggerName = logEvent.getLoggerName();
        String message = new String(this.getLayout().toByteArray(logEvent));
        Date eventTime = new Date(logEvent.getTimeMillis());

        Map<String, Object> m = new HashMap<>();
        m.put("level", level);

        StackTraceElement logSource = logEvent.getSource();

        if (logSource == null || logSource.getFileName() == null) {
            m.put("sourceFile", "<unknown>");
        } else {
            m.put("sourceFile", logSource.getFileName());
        }

        if (logSource == null || logSource.getClassName() == null) {
            m.put("sourceClass", "<unknown>");
        } else {
            m.put("sourceClass", logEvent.getSource().getClassName());
        }

        if (logSource == null || logSource.getMethodName() == null) {
            m.put("sourceMethod", "<unknown>");
        } else {
            m.put("sourceMethod", logEvent.getSource().getMethodName());
        }

        if (logSource == null || logSource.getLineNumber() == 0) {
            m.put("sourceLine", 0);
        } else {
            m.put("sourceLine", logEvent.getSource().getLineNumber());
        }

        try {
            if(usePre26Abbreviate) {
                m.put("logger", abbreviatePre26(loggerName));
            } else if(abbreviateMethod != null) {
                m.put("logger", abbreviate(loggerName));
            } else {
                //just a safety net in case abbreviate() changes again in a future API
                m.put("logger", loggerName);
            }
        } catch (final IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
            LOG.error(e.getMessage());
        }
        m.put("loggerFull", loggerName);
        m.put("message", message);
        m.put("thread", logEvent.getThreadName());
        m.putAll(this.staticFields);

        // TODO: get rid of that, once the whole stack supports subsecond timestamps
        // this is just a workaround due to the lack of support
        m.put("@timestamp", format.format(eventTime));

        if (this.fluency != null) {
            try {
                // the tag is required for further processing within fluentd,
                // otherwise we would have no way to manipulate messages in transit
                this.fluency.emit((String) parameters.get("tag"), logEvent.getTimeMillis(), m);
            } catch (IOException e) {
                LOG.error(e.getMessage());
            }
        }
    }

    /**
     * Starting with Log4J API 2.6 the abbreviate() method signature changed. This method works with API versions 2.5 or older.
     * 
     * @param stringToAbbreviate
     * @return
     * @throws InvocationTargetException 
     * @throws IllegalArgumentException 
     * @throws IllegalAccessException 
     */
    private String abbreviatePre26(final String stringToAbbreviate) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
        return (String) abbreviateMethod.invoke(abbr, stringToAbbreviate);
    }
    
    /**
     * Starting with Log4J API 2.6 the abbreviate() method signature changed. This method works with API versions 2.6 or newer.
     * 
     * @param stringToAbbreviate
     * @return
     * @throws InvocationTargetException 
     * @throws IllegalArgumentException 
     * @throws IllegalAccessException 
     */
    private String abbreviate(final String stringToAbbreviate) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {
        final StringBuilder logger = new StringBuilder();
        abbreviateMethod.invoke(abbr, stringToAbbreviate, logger);
        return logger.toString();
    }
    
}
